h/models/group.py

```diff
@@ -159,12 +166,31 @@ def groupid(self, value):
             self.authority_provided_id = groupid_parts["authority_provided_id"]
             self.authority = groupid_parts["authority"]
 
-    # Group membership
-    members = sa.orm.relationship(
-        "User",
-        secondary="user_group",
-        backref=sa.orm.backref("groups", order_by="Group.name"),
-    )
+    memberships = sa.orm.relationship("GroupMembership", back_populates="group")
+
+    @property
+    def members(self) -> tuple[User, ...]:
+        """
+        Return a tuple of this group's members.
+
+        This is a convenience property for when you want to access a group's
+        members (User objects) rather than its memberships (GroupMembership
+        objects).
+
+        This is not an SQLAlchemy relationship! SQLAlchemy emits a warning if
+        you try to have both Group.memberships and a Group.members
+        relationships at the same time because it can result in reads returning
+        conflicting data and in writes causing integrity errors or unexpected
+        inserts or deletes. See:
+
+        https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#combining-association-object-with-many-to-many-access-patterns
+
+        Since this is just a normal Python property setting or mutating it
+        (e.g. `group.members = [...]` or `group.members.append(...)`) wouldn't
+        be registered with SQLAlchemy and the changes wouldn't be saved to the
+        DB. So this is a read-only property that returns an immutable tuple.
+        """
+        return tuple(membership.user for membership in self.memberships)
```

Comment:
This would make a query per user in the group if I'm not mistaken.

Reply:
Yeah, it makes one `SELECT user_group` and then three `SELECT user`'s. I think it might be worth fixing that, I'll have a look into SQLAlchemy's relationship loading stuff...

Reply:
Great spot :+1: For now I've changed the `GroupMembership.user` and `GroupMembership.group` relationships to use eager loading by default, which avoids the n+1 problem in this line. I think eager loading is probably a better default for these relationships because with lazy loading it's _very_ easy to write code that introduces the n+1 `SELECT`'s problem and not realise it. [Slack thread](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1731067621404349)